subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/services/DefaultEnvironmentChangeTracker.kt

```diff
@@ -105,19 +119,47 @@ class DefaultEnvironmentChangeTracker(private val problemFactory: ProblemFactory
         }
 
         fun isSystemPropertyMutated(key: String): Boolean {
-            return systemPropertiesCleared || mutatedSystemProperties.containsKey(key)
+            return systemPropertiesCleared || mutatedSystemProperties[key] is SystemPropertyMutate
+        }
+
+        fun isSystemPropertyLoaded(key: String): Boolean {
+            return mutatedSystemProperties[key] is SystemPropertyLoad
+        }
+
+        fun getLoadedPropertyOldValue(key: String): Any? {
+            return (mutatedSystemProperties[key] as? SystemPropertyLoad)?.oldValue
         }
 
         fun getCachedState(): CachedEnvironmentState {
             return CachedEnvironmentState(
                 cleared = systemPropertiesCleared,
-                updates = mutatedSystemProperties.values.filterIsInstance<SystemPropertySet>(),
+                updates = mutatedSystemProperties.values
+                    .filterIsInstance<SystemPropertySet>()
+                    .filter { it !is SystemPropertyOverride },
                 removals = mutatedSystemProperties.values.filterIsInstance<SystemPropertyRemove>()
             )
         }
 
+        fun systemPropertyOverridden(key: Any, value: Any?) {
+            mutatedSystemProperties[key] = SystemPropertyOverride(key, value)
+        }
+
+        fun systemPropertyLoadedByRootBuild(key: Any, value: Any?) {
+            mutatedSystemProperties[key] = SystemPropertyRootLoad(key, value)
+        }
+
+        fun systemPropertyLoaded(key: Any, value: Any?, oldValue: Any?) {
+            val loadedOldValue = when (val loadedSystemProperty = mutatedSystemProperties[key]) {
+                is SystemPropertyRootLoad -> loadedSystemProperty.value
```

Comment:
Isn't `oldValue` always equal to the `loadedSystemProperty.value` in this case? I suppose, system properties from root build are restored outside this class, so we don't have to track system properties in this class as well - we just must not store such system properties as `SystemPropertyLoad`.